verl/utils/rollout_trace.py

```diff
@@ -0,0 +1,192 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+import contextlib
+import functools
+import inspect
+from typing import Optional
+
+
+class RolloutTraceConfig:
+    _instance: Optional["RolloutTraceConfig"] = None
+    backend: Optional[str] = None
+    client: Optional[object] = None
+    token2text: bool = False
+
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    @classmethod
+    def get_instance(cls) -> "RolloutTraceConfig":
+        if cls._instance is None:
+            cls._instance = cls()
+        return cls._instance
+
+    @classmethod
+    def init(cls, project_name: str, experiment_name: str, backend: str, token2text: bool = False):
+        config = cls.get_instance()
+        config.backend = backend
+        config.token2text = token2text
+        config.project_name = project_name
+        config.experiment_name = experiment_name
+        if backend == "weave":
+            import weave
+
+            config.client = weave.init(project_name)
+        elif backend == "mlflow":
+            import mlflow
+
+            config.client = mlflow
+        else:
+            config.client = None
+
+    @classmethod
+    def get_backend(cls) -> Optional[str]:
+        return cls.get_instance().backend
+
+    @classmethod
+    def get_client(cls) -> Optional[object]:
+        return cls.get_instance().client
+
+    @classmethod
+    def enable_token2text(cls) -> Optional[bool]:
+        return cls.get_instance().token2text
+
+    @classmethod
+    def reset(cls):
+        cls._instance = None
+
+
+@contextlib.contextmanager
+def rollout_trace_attr(sample_index=None, step=None, rollout_n=None):
+    """A context manager to add attributes to a trace for the configured backend."""
+    backend = RolloutTraceConfig.get_backend()
+    attributes = {}
+    if sample_index is not None:
+        attributes["sample_index"] = sample_index
+    if step is not None:
+        attributes["step"] = step
+    if rollout_n is not None:
+        attributes["rollout_n"] = rollout_n
+    attributes["experiment_name"] = RolloutTraceConfig.get_instance().experiment_name
+
+    if not attributes or backend is None:
+        yield
+        return
+
+    if backend == "weave":
+        import weave
+
+        with weave.attributes(attributes):
+            yield
+    # TODO implement mlfow trace
+    # elif backend == "mlflow":
+    #     import mlflow
+    #     # This assumes a run is already active.
+    #     # We are setting tags for the current active run.
+    #     try:
+    #         mlflow.set_tags(attributes)
+    #     except Exception:
+    #         # Silently fail if there is no active run.
+    #         pass
+    #     yield
+    else:
+        yield
+
+
+def rollout_trace_op(func):
```

Comment:
Seems like this function can trace every function, include tooling, agent loop run, ...?